File: problems/04_stack/0739_daily_temperatures/solution.swift
// Problem: 0739. Daily Temperatures
// Difficulty: Medium
// URL: https://leetcode.com/problems/daily-temperatures/
//
// 最佳解：單調遞減堆疊。

import Foundation

class Solution {
    func dailyTemperatures(_ temperatures: [Int]) -> [Int] {
        let n = temperatures.count
        var ans = Array(repeating: 0, count: n)
        var stack: [Int] = []
        for i in 0..<n {
            while let top = stack.last, temperatures[top] < temperatures[i] {
                stack.removeLast()
                ans[top] = i - top
            }
            stack.append(i)
        }
        return ans
    }

    func dailyTemperatures_brute(_ temperatures: [Int]) -> [Int] {
        let n = temperatures.count
        var ans = Array(repeating: 0, count: n)
        for i in 0..<n {
            for j in (i + 1)..<n {
                if temperatures[j] > temperatures[i] {
                    ans[i] = j - i
                    break
                }
            }
        }
        return ans
    }
}

func checkArr(_ a: [Int], _ e: [Int], _ msg: String) {
    if a != e { fatalError("FAIL \(msg): \(a) vs \(e)") }
}

let sol = Solution()
for fn in [sol.dailyTemperatures, sol.dailyTemperatures_brute] {
    checkArr(fn([73, 74, 75, 71, 69, 72, 76, 73]), [1, 1, 4, 2, 1, 1, 0, 0], "ex1")
    checkArr(fn([30, 40, 50, 60]), [1, 1, 1, 0], "ex2")
    checkArr(fn([30, 60, 90]), [1, 1, 0], "ex3")
    checkArr(fn([90, 80, 70]), [0, 0, 0], "desc")
    checkArr(fn([50]), [0], "single")
    checkArr(fn([55, 55, 55, 60]), [3, 2, 1, 0], "dup")
}
print("All tests passed!")

File: problems/04_stack/0739_daily_temperatures/solution.dart
// Problem: 0739. Daily Temperatures
// Difficulty: Medium
// URL: https://leetcode.com/problems/daily-temperatures/
//
// 最佳解：單調遞減堆疊。O(n) / O(n)。

List<int> dailyTemperatures(List<int> temperatures) {
  final n = temperatures.length;
  final ans = List<int>.filled(n, 0);
  final stack = <int>[];
  for (int i = 0; i < n; i++) {
    while (stack.isNotEmpty && temperatures[stack.last] < temperatures[i]) {
      final j = stack.removeLast();
      ans[j] = i - j;
    }
    stack.add(i);
  }
  return ans;
}

List<int> dailyTemperaturesBrute(List<int> temperatures) {
  final n = temperatures.length;
  final ans = List<int>.filled(n, 0);
  for (int i = 0; i < n; i++) {
    for (int j = i + 1; j < n; j++) {
      if (temperatures[j] > temperatures[i]) {
        ans[i] = j - i;
        break;
      }
    }
  }
  return ans;
}

void _expect(List<int> a, List<int> e, String msg) {
  if (a.length != e.length) throw StateError('FAIL $msg: len');
  for (int i = 0; i < a.length; i++) {
    if (a[i] != e[i]) throw StateError('FAIL $msg @$i: $a vs $e');
  }
}

void main() {
  for (final fn in [dailyTemperatures, dailyTemperaturesBrute]) {
    _expect(fn([73, 74, 75, 71, 69, 72, 76, 73]), [1, 1, 4, 2, 1, 1, 0, 0], 'ex1');
    _expect(fn([30, 40, 50, 60]), [1, 1, 1, 0], 'ex2');
    _expect(fn([30, 60, 90]), [1, 1, 0], 'ex3');
    _expect(fn([90, 80, 70]), [0, 0, 0], 'desc');
    _expect(fn([50]), [0], 'single');
    _expect(fn([55, 55, 55, 60]), [3, 2, 1, 0], 'dup');
  }
  print('All tests passed!');
}
